### PR TITLE
agent-ctl: Allow API specification in JSON format

### PR DESCRIPTION
**Pull request by jodh-intel**
Update the `agent-ctl` tool to allow API fields to be specified in JSON
format, either directly on the command-line, or via a file URI.

This feature is made possible by enabling `serde` support in the agent
`protocols` crate. Careful use of the `serde` macros allows the
`agent-ctl` tool to accept _partially_ specified API objects in JSON
format; fields that are not specified are set to the default value for
their respective types.

`build.rs` changes based on work by Fupan.

Fixes: #2978.

Contributions-by: Fupan Li <lifupan@gmail.com>
Contributions-by: Bin Liu <bin@hyper.sh>
Signed-off-by: James O. D. Hunt <james.o.hunt@intel.com>

### CODE REVIEW

**Comment by jodh-intel** [comment]
@lifupan, @liubin - could you tal?

**Comment by jodh-intel** [comment]
The `jenkins-vfio-host:ubuntu18-04-vm:fedora32` is failing consistently like this:

```
10:13:13 Fedora 32 - x86_64 - Updates                    6.6 MB/s |  30 MB     00:04    
10:28:13 Build timed out (after 15 minutes). Marking the build as aborted.
```

Could this be due to F32 being EOL now I wonder?

**Comment by fidencio** [comment]
> The `jenkins-vfio-host:ubuntu18-04-vm:fedora32` is failing consistently like this:
> 
> ```
> 10:13:13 Fedora 32 - x86_64 - Updates                    6.6 MB/s |  30 MB     00:04    
> 10:28:13 Build timed out (after 15 minutes). Marking the build as aborted.
> ```
> 
> Could this be due to F32 being EOL now I wonder?

Hmm, weird.  Fedora 32 is EOL since May (https://lists.fedoraproject.org/archives/list/announce@lists.fedoraproject.org/thread/RPZQXPEV3L45YQV2HRBIQ5DQZ7GMP4X3/).

**Comment by jodh-intel** [comment]
I guess that's not the reason then. I'll try once more...

**Comment by jodh-intel** [comment]
Looks like either the CI had network issues or the mirror did as it's now working.

File: src/agent/protocols/build.rs [hunk 1]
```diff
@@ -3,29 +3,152 @@
...
+    dst.write_all(new_contents.as_bytes())?;
+
+    Ok(())
+}
+
+fn use_serde(protos: &[&str], out_dir: &Path) -> Result<(), std::io::Error> {
+    protos
+        .iter()
+        .try_for_each(|f: &&str| -> Result<(), std::io::Error> {
+            let proto_filename = Path::new(f)
```
Comment (liubin) [inline src/agent/protocols/build.rs:30]: These few lines are a bit complicated, how about 
```rust
    let out_file = Path::new(f)
        .file_name()
        .and_then(|s| s.to_str())
        .ok_or("failed to get proto file name")
        .map(|s| {
            let t = s.replace(".proto", ".rs");
            out_dir.join(t)
        })
        .map_err(|e| std::io::Error::new(std::io::ErrorKind::Other, e))?;
```

Comment (jodh-intel) [reply src/agent/protocols/build.rs]: Thanks @liubin - I've added that code and also added you as a contributor. Note that I had to tweak it a bit to convert the `PathBuf` back into a string for the call to `replace_text_in_file()`.

**Comment by jodh-intel** [comment]
Any rustaceans out there who could help review this agent control tool change?

**Comment by jodh-intel** [comment]
@bergwolf - could you tal?

**Comment by jodh-intel** [comment]
Grrr, looks like the Fedora vfio CI has got "stuck".

/retest-vfio

**Comment by jodh-intel** [comment]
CI is all green - can we get some reviews on this please?

**Review by fidencio — APPROVED** [review]
@jodh-intel, lgtm!

Please, take this with a grain of salt due to my very very basic rust skills.  You may want a review from someone else more experienced with the language than myself.

**Review by liubin — APPROVED** [review]

File: src/agent/protocols/build.rs [hunk 2]
```diff
@@ -3,29 +3,152 @@
...
+
+    let mut dst = File::create(&rust_filename)?;
+
+    dst.write_all(data.as_bytes())?;
+
+    Ok(())
+}
+
+fn real_main() -> Result<(), std::io::Error> {
+    let autogen_comment = format!("\n//! Generated by {:?} ({:?})", file!(), module_path!());
```
[the remark this answers is not in the document]
Comment (jodh-intel) [reply src/agent/protocols/build.rs:94]: Yes, makes sense. I'd do this on the next update.

**Pull request merged**